SlaveWAMPServer.js

```diff
@@ -56,151 +39,58 @@ class SlaveWAMPServer extends WAMPServer {
 				}
 				configuredCb(null, this);
 				configuredCb = () => {};
+			} else {
+				console.error(`Received invalid master message payload of type "${payload.type}"`);
 			}
 		});
 	}
 
-	/**
-	 * @param {Object}[request={}] request
-	 * @returns {RPCRequestSchema}
-	 */
-	static normalizeRequest(request = {}) {
-		if (!request.procedure || typeof request.procedure !== 'string') {
-			throw new Error(`Wrong format of requested procedure: ${request.procedure}`);
-		}
-		if (!request.socketId || typeof request.socketId !== 'string') {
-			throw new Error(`Wrong format of requested socket id: ${request.socketId}`);
-		}
-		request.procedure = request.procedure.replace(/\./g, '');
-		request.socketId = request.socketId.replace(/\./g, '');
-		return request;
-	}
-
 	/**
 	 * @param {string} procedure
 	 * @param {*} data
-	 * @param {string} socketId
-	 * @param {Function} cb
+	 * @param {Function} callback
 	 * @returns {undefined}
 	 */
-	sendToMaster(procedure, data, socketId, cb) {
-		const req = SlaveWAMPServer.normalizeRequest({
+	sendToMaster(procedure, data, callback) {
+		const req = {
 			type: schemas.InterProcessRPCRequestSchema.id,
 			procedure,
 			data,
-			socketId,
-			workerId: this.worker.id,
-			signature: WAMPClient.generateSignature(get(this.interProcessRPC, `${socketId}.${procedure}`, {})),
-		});
-		this.worker.sendToMaster(req);
-
-		this.saveCall(req, cb);
+		};
+		if (callback) {
+			this.worker.sendToMaster(req, (err, response) => {
+				if (err) {
+					return callback(err);
+				}
+				return callback(null, response.data);
+			});
+		} else {
+			this.worker.sendToMaster(req);
+		}
 	}
 
 	/**
 	 * @param {RPCRequestSchema} request
-	 * @param {Object} socket
+	 * @param {Object} respond
 	 * @returns {undefined}
 	 */
-	processWAMPRequest(request, socket) {
+	processWAMPRequest(request, respond) {
 		if (v.validate(request, schemas.RPCRequestSchema).valid) {
-			request.socketId = socket.id;
-			request.workerId = this.worker.id;
 			if (this.endpoints.slaveRpc[request.procedure] &&
 				typeof this.endpoints.slaveRpc[request.procedure] === 'function') {
-				this.endpoints.slaveRpc[request.procedure](request,
-					this.reply.bind(this, socket, request));
+				this.endpoints.slaveRpc[request.procedure](request, (error, data) => {
+					respond(error, {
+						type: schemas.RPCResponseSchema.id,
+						data,
+					});
+				});
 			} else {
 				request.type = schemas.MasterRPCRequestSchema.id;
-				this.worker.sendToMaster(request);
+				this.worker.sendToMaster(request, respond);
```

Comment:
Let's reuse sendToMaster defined in SlaveWAMPServer that handles callbacks in the correct way

Reply:
Actually maybe we should keep it as is because the `sendToMaster` which we expose directly has a different request id/type - To distinguish between a request which originates from a client versus a request which originates from the server.